trace_decoder/src/imp.rs

```diff
@@ -0,0 +1,641 @@
+            state: SeparateTriePreImage::Direct(state),
+            storage: SeparateStorageTriesPreImage::MultipleTries(storage),
+        }) => {
+            let state = state.items().try_fold(
+                StateMpt::default(),
+                |mut acc, (nibbles, hash_or_val)| {
+                    let path = TrieKey::from_nibbles(nibbles);
+                    match hash_or_val {
+                        mpt_trie::trie_ops::ValOrHash::Val(bytes) => {
+                            #[expect(deprecated)] // this is MPT specific
+                            acc.insert_by_hashed_address(
+                                path.into_hash()
+                                    .context("invalid path length in direct state trie")?,
+                                rlp::decode(&bytes)
+                                    .context("invalid AccountRlp in direct state trie")?,
+                            )?;
+                        }
+                        mpt_trie::trie_ops::ValOrHash::Hash(h) => {
+                            acc.insert_hash_by_key(path, h)?;
+                        }
+                    };
+                    anyhow::Ok(acc)
+                },
+            )?;
+            let storage = storage
+                .into_iter()
+                .map(|(k, SeparateTriePreImage::Direct(v))| {
+                    v.items()
+                        .try_fold(StorageTrie::default(), |mut acc, (nibbles, hash_or_val)| {
+                            let path = TrieKey::from_nibbles(nibbles);
+                            match hash_or_val {
+                                mpt_trie::trie_ops::ValOrHash::Val(value) => {
+                                    acc.insert(path, value)?;
+                                }
+                                mpt_trie::trie_ops::ValOrHash::Hash(h) => {
+                                    acc.insert_hash(path, h)?;
+                                }
+                            };
+                            anyhow::Ok(acc)
+                        })
+                        .map(|v| (k, v))
+                })
+                .collect::<Result<_, _>>()?;
+            (state, storage, Hash2Code::new())
+        }
+        BlockTraceTriePreImages::Combined(CombinedPreImages { compact }) => {
+            let instructions = crate::wire::parse(&compact)
+                .context("couldn't parse instructions from binary format")?;
+            let crate::type1::Frontend {
+                state,
+                storage,
+                code,
+            } = crate::type1::frontend(instructions)?;
+            (state, storage, code.into_iter().map(Into::into).collect())
+        }
+    })
+}
+
+/// Break `txns` into batches of length `hint`, prioritising creating at least
+/// two batches.
+///
+/// [`None`] represents a dummy transaction that should not increment the
+/// transaction index.
+fn batch(txns: Vec<TxnInfo>, hint: usize) -> Vec<Vec<Option<TxnInfo>>> {
```

Comment:
why not keeping `batch_size` as var name here? Which is much more straightforward for less familiar readers diving into the code and possibly skipping the rustdoc